File: sample-videochat-webrtc/sample-videochat-webrtc/Components/Users/Views/SelectedUsersView/SelectedUsersView.m
//
//  UserTagView.m
//  sample-videochat-webrtc
//
//  Created by Injoit on 03.08.2021.
//  Copyright © 2021 QuickBlox Team. All rights reserved.
//

#import "SelectedUsersView.h"
#import "UserTagView.h"

const CGFloat paddingLeft = 10.0f;
const CGFloat spaceBetween = 2.0f;
const CGFloat heightView = 24.0f;

@interface SelectedUsersView()
//MARK: - IBOutlets
@property (weak, nonatomic) IBOutlet UILabel *maxUsersLabel;
//MARK: - Properties
@property (strong, nonatomic) NSMutableArray<UserTagView *> *selectedViews;
@property (strong, nonatomic) NSLayoutConstraint *topConstraint;
@end

@implementation SelectedUsersView
//MARK: - Life Cycle
- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        self.selectedViews = @[].mutableCopy;
    }
    return self;
}

//MARK: - Public Methods
- (void)addViewWithUserID:(NSUInteger)userID userName:(NSString *)userName {
    UserTagView *selectedUserView = [[NSBundle mainBundle] loadNibNamed:@"UserTagView"
                                                                  owner:nil
                                                                options:nil].firstObject;
    selectedUserView.name = userName;
    selectedUserView.userID = @(userID);
    [self addSubview:selectedUserView];
    [self.selectedViews addObject:selectedUserView];
    self.maxUsersLabel.hidden = YES;
    [self setupViews];
    
    __weak __typeof(self)weakSelf = self;
    [selectedUserView setOnCancelTapped:^(NSUInteger userID) {
        [weakSelf removeViewWithUserID:userID];
        if (weakSelf.onSelectedUserViewCancelTapped) {
            weakSelf.onSelectedUserViewCancelTapped(userID);
        }
        [weakSelf removeViewWithUserID:userID];
    }];
}

- (void)removeViewWithUserID:(NSUInteger)userID {
    NSPredicate *predicate = [NSPredicate predicateWithFormat:@"userID == %@", @(userID)];
    UserTagView *selectedUserView = [self.selectedViews filteredArrayUsingPredicate:predicate].firstObject;
    if (!selectedUserView) {
        return;
    }
    NSPredicate *removePredicate = [NSPredicate predicateWithFormat:@"userID != %@", @(userID)];
    self.selectedViews = [self.selectedViews.copy filteredArrayUsingPredicate:removePredicate].mutableCopy;
    
    [selectedUserView removeFromSuperview];
    if (self.selectedViews.count == 0) {
        self.maxUsersLabel.hidden = NO;
        return;
    }
    self.maxUsersLabel.hidden = YES;
    for (UserTagView *view in self.selectedViews) {
        [view removeFromSuperview];
        [self addSubview:view];
    }
    [self setupViews];
}

- (void)clear {
    for (UserTagView *view in self.selectedViews) {
        [view removeFromSuperview];
    }
    self.selectedViews = @[].mutableCopy;
    self.maxUsersLabel.hidden = NO;
}

//MARK - Setup
- (void)setupViews {
    UserTagView *previousView = nil;
    CGFloat viewsWidth = 0.0f;
    NSUInteger spaceCount = 0;
    
    for (int i = 0; i < self.selectedViews.count; i++) {
        UserTagView *selectedUserView = self.selectedViews[i];
        selectedUserView.translatesAutoresizingMaskIntoConstraints = NO;
        [selectedUserView.heightAnchor constraintEqualToConstant:heightView].active = YES;
        CGFloat selectedUserViewWidth = selectedUserView.nameLabel.intrinsicContentSize.width + 37.0f;
        [selectedUserView.widthAnchor constraintEqualToConstant:selectedUserViewWidth].active = YES;
        if (previousView) {
            CGFloat allwidth = paddingLeft + (spaceBetween * spaceCount) + viewsWidth + selectedUserViewWidth;
            if (self.bounds.size.width > allwidth) {
                spaceCount = spaceCount + 1;
                [selectedUserView.leftAnchor constraintEqualToAnchor:previousView.rightAnchor constant:spaceBetween].active = YES;
                [selectedUserView.topAnchor constraintEqualToAnchor:previousView.topAnchor].active = YES;
            } else {
                self.topConstraint.constant = 3.0f;
                spaceCount = 0;
                viewsWidth = 0.0f;
                [selectedUserView.leftAnchor constraintEqualToAnchor:self.leftAnchor constant:paddingLeft].active = YES;
                [selectedUserView.topAnchor constraintEqualToAnchor:previousView.bottomAnchor constant:3.0f].active = YES;
            }
        } else {
            spaceCount = 1;
            self.topConstraint = [selectedUserView.topAnchor constraintEqualToAnchor:self.topAnchor constant:18.0f];
            self.topConstraint.active = YES;
            [selectedUserView.leftAnchor constraintEqualToAnchor:self.leftAnchor constant:paddingLeft].active = YES;
        }
        previousView = selectedUserView;
        viewsWidth = viewsWidth + selectedUserViewWidth;
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/Components/Users/Views/SelectedUsersView/SelectedUsersView.swift
//
//  SelectedUsersView.swift
//  sample-videochat-webrtc-swift
//
//  Created by Injoit on 29.07.2021.
//  Copyright © 2021 QuickBlox. All rights reserved.
//

import UIKit

typealias SelectUserViewCancelCompletion = (_ userID: UInt) -> Void

struct SelectedUsersViewConstants {
    static let paddingLeft: CGFloat = 14.0
    static let spaceBetween: CGFloat = 2.0
    static let heightView: CGFloat = 24.0
}

class SelectedUsersView: UIView {
    //MARK: - IBOutlets
    @IBOutlet weak var maxUsersLabel: UILabel!
    @IBOutlet weak var containerView: UIView!
    
    //MARK: - Properties
    var onSelectedUserViewCancelTapped: SelectUserViewCancelCompletion?
    private var topConstraint: NSLayoutConstraint!
    private var selectedViews: [SelectedUserView] = [] {
        didSet {
            maxUsersLabel.isHidden = !selectedViews.isEmpty
        }
    }
    
    //MARK: - Public Methods
    func addView(_ userID: UInt, userName: String = "User") {
        let selectedUserView = SelectedUserView.loadNib()
        selectedUserView.name = userName
        selectedUserView.userID = userID
        
        selectedUserView.onCancelTapped = { [weak self] (userID) in
            self?.removeView(userID)
            self?.onSelectedUserViewCancelTapped?(userID)
        }
        
        containerView.addSubview(selectedUserView)
        selectedViews.append(selectedUserView)
        setupViews()
    }
    
    func removeView(_ userID: UInt) {
        guard let selectedView = self.selectedViews.filter({ $0.userID == userID }).first,
              let index = self.selectedViews.firstIndex(of: selectedView) else {
            return
        }
        selectedViews.remove(at: index)
        selectedView.removeFromSuperview()
        if selectedViews.count == 0 {
            return
        }
        for view in selectedViews {
            view.removeFromSuperview()
            containerView.addSubview(view)
        }
        setupViews()
    }
    
    func clear() {
        for selectedUserView in selectedViews {
            selectedUserView.removeFromSuperview()
        }
        selectedViews = []
    }
    
    //MARK: - Private Methods
    private func setupViews() {
        var previousView: SelectedUserView? = nil
        var viewsWidth: CGFloat = 0.0
        var spaceCount = 0
        for i in 0...selectedViews.count - 1 {
            let selectedUserView = selectedViews[i]
            selectedUserView.translatesAutoresizingMaskIntoConstraints = false
            selectedUserView.heightAnchor.constraint(equalToConstant: SelectedUsersViewConstants.heightView).isActive = true
            let selectedUserViewWidth: CGFloat = selectedUserView.nameLabel.intrinsicContentSize.width + 37.0
            selectedUserView.widthAnchor.constraint(equalToConstant: selectedUserViewWidth).isActive = true
            if let previous = previousView {
                let allwidth = SelectedUsersViewConstants.paddingLeft + (SelectedUsersViewConstants.spaceBetween * CGFloat(spaceCount)) + viewsWidth + selectedUserViewWidth
                if bounds.width > allwidth {
                    spaceCount = spaceCount + 1
                    selectedUserView.leftAnchor.constraint(equalTo: previous.rightAnchor, constant: SelectedUsersViewConstants.spaceBetween).isActive = true
                    selectedUserView.topAnchor.constraint(equalTo: previous.topAnchor).isActive = true
                } else {
                    topConstraint.constant = 3.0
                    spaceCount = 0
                    viewsWidth = 0.0
                    selectedUserView.leftAnchor.constraint(equalTo: leftAnchor, constant: SelectedUsersViewConstants.paddingLeft).isActive = true
                    selectedUserView.topAnchor.constraint(equalTo: previous.bottomAnchor, constant: 3.0).isActive = true
                }
            } else {
                spaceCount = 1
                topConstraint = selectedUserView.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 18.0)
                topConstraint.isActive = true
                selectedUserView.leftAnchor.constraint(equalTo: leftAnchor, constant: SelectedUsersViewConstants.paddingLeft).isActive = true
            }
            previousView = selectedUserView
            viewsWidth = viewsWidth + selectedUserViewWidth
        }
    }
}
